File: app/src/main/java/com/metahumanz/pacilread/AppDrawerController.java
package com.metahumanz.pacilread;

import android.app.Activity;
import android.graphics.Rect;
import android.os.Build;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewConfiguration;
import android.view.ViewGroup;
import android.view.animation.DecelerateInterpolator;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;

public class AppDrawerController {
    private static final float DRAWER_WIDTH_RATIO = 0.84f;
    private static final int DRAWER_MIN_WIDTH_DP = 300;
    private static final int DRAWER_MAX_WIDTH_DP = 420;
    private static final int EDGE_ACTIVATION_MIN_DP = 112;
    private static final int EDGE_ACTIVATION_MAX_DP = 156;
    private static final int EDGE_EXCLUSION_DP = 156;
    private static final int OPEN_DRAWER_DRAG_EXTRA_DP = 72;
    private static final float OPEN_THRESHOLD_RATIO = 0.34f;
    private static final float CLOSE_THRESHOLD_RATIO = 0.22f;
    private static final float FLING_VELOCITY_THRESHOLD = 0.55f;
    public static final int SECTION_NONE = -1;
    public static final int SECTION_BOOKSHELF = 0;
    public static final int SECTION_PREVIEW = 1;
    public static final int SECTION_SETTINGS = 2;

    public interface NavigationListener {
        void onDrawerDestinationSelected(int destination);
    }

    private final Activity activity;
    private final View rootView;
    private final View drawerPanel;
    private final View drawerScrim;
    private final View navBookshelf;
    private final View navPreview;
    private final View navSettings;
    private final TextView navBookshelfText;
    private final TextView navPreviewText;
    private final TextView navSettingsText;
    private final TextView statusText;
    private final NavigationListener navigationListener;
    private final int touchSlop;

    private int currentSection = SECTION_NONE;
    private boolean drawerOpen = false;
    private boolean drawerGestureCandidate = false;
    private boolean drawerDragging = false;
    private float drawerDownX = 0f;
    private float drawerDownY = 0f;
    private float drawerBaseOffset = 0f;
    private float drawerLastX = 0f;
    private long drawerLastEventTime = 0L;
    private float drawerVelocityX = 0f;
    private boolean pendingChildTouchCancel = false;
    private boolean drawerGestureStartedOpen = false;
    private long drawerAnimationToken = 0L;

    public AppDrawerController(Activity activity, View rootView, NavigationListener navigationListener) {
        this.activity = activity;
        this.rootView = rootView;
        this.navigationListener = navigationListener;
        this.drawerPanel = activity.findViewById(R.id.drawer_panel);
        this.drawerScrim = activity.findViewById(R.id.drawer_scrim);
        this.navBookshelf = activity.findViewById(R.id.nav_bookshelf);
        this.navPreview = activity.findViewById(R.id.nav_preview);
        this.navSettings = activity.findViewById(R.id.nav_settings);
        this.navBookshelfText = activity.findViewById(R.id.text_nav_bookshelf);
        this.navPreviewText = activity.findViewById(R.id.text_nav_preview);
        this.navSettingsText = activity.findViewById(R.id.text_nav_settings);
        this.statusText = activity.findViewById(R.id.text_drawer_status);
        this.touchSlop = ViewConfiguration.get(activity).getScaledTouchSlop();
        setupViews();
    }

    private void setupViews() {
        if (drawerPanel == null || drawerScrim == null) {
            return;
        }
        drawerScrim.setOnClickListener(v -> closeDrawer());
        navBookshelf.setOnClickListener(v -> {
            navigationListener.onDrawerDestinationSelected(SECTION_BOOKSHELF);
            closeDrawer();
        });
        navPreview.setOnClickListener(v -> {
            navigationListener.onDrawerDestinationSelected(SECTION_PREVIEW);
            closeDrawer();
        });
        navSettings.setOnClickListener(v -> {
            navigationListener.onDrawerDestinationSelected(SECTION_SETTINGS);
            closeDrawer();
        });

        drawerPanel.post(() -> {
            updateDrawerPanelWidth();
            setDrawerOffset(-drawerPanel.getWidth());
            drawerPanel.setVisibility(View.INVISIBLE);
            drawerScrim.setVisibility(View.GONE);
            drawerScrim.setAlpha(0f);
        });
        if (rootView != null) {
            rootView.post(() -> {
                updateDrawerPanelWidth();
                syncDrawerOffsetToCurrentState();
                updateDrawerGestureExclusion();
            });
            rootView.addOnLayoutChangeListener((view, left, top, right, bottom, oldLeft, oldTop, oldRight, oldBottom) ->
                    handleRootLayoutChanged()
            );
        }
        updateNavigationState();
    }

    public void bindMenuButton(int viewId) {
        View button = activity.findViewById(viewId);
        if (button != null) {
            button.setOnClickListener(v -> openDrawer());
        }
    }

    public void setCurrentSection(int section) {
        currentSection = section;
        updateNavigationState();
    }

    public void setStatusText(String text) {
        if (statusText != null) {
            statusText.setText(text == null ? "" : text);
        }
    }

    public boolean onBackPressed() {
        if (isDrawerVisible()) {
            closeDrawer();
            return true;
        }
        return false;
    }

    public boolean handleTouchEvent(MotionEvent event) {
        if (drawerPanel == null || drawerPanel.getWidth() == 0) {
            return false;
        }
        float x = event.getX();
        float y = event.getY();
        switch (event.getActionMasked()) {
            case MotionEvent.ACTION_DOWN:
                pendingChildTouchCancel = false;
                drawerGestureCandidate = shouldStartDrawerGesture(x);
                drawerDragging = false;
                drawerVelocityX = 0f;
                drawerGestureStartedOpen = isDrawerVisible();
                if (drawerGestureCandidate) {
                    drawerDownX = x;
                    drawerDownY = y;
                    drawerLastX = x;
                    drawerLastEventTime = event.getEventTime();
                    drawerBaseOffset = drawerPanel.getTranslationX();
                }
                return false;
            case MotionEvent.ACTION_MOVE:
                if (!drawerGestureCandidate) {
                    return false;
                }
                float deltaX = x - drawerDownX;
                float deltaY = y - drawerDownY;
                if (!drawerDragging) {
                    int horizontalThreshold = isDrawerVisible() ? touchSlop : Math.max(4, touchSlop / 2);
                    if (Math.abs(deltaY) > touchSlop && Math.abs(deltaY) > Math.abs(deltaX)) {
                        drawerGestureCandidate = false;
                        return false;
                    }
                    if (Math.abs(deltaX) <= horizontalThreshold || Math.abs(deltaX) <= Math.abs(deltaY)) {
                        return false;
                    }
                    if (!isDrawerVisible() && deltaX < 0f) {
                        drawerGestureCandidate = false;
                        return false;
                    }
                    drawerDragging = true;
                    pendingChildTouchCancel = true;
                    prepareDrawerForInteraction(false);
                }
                float targetOffset = clamp(drawerBaseOffset + deltaX, -drawerPanel.getWidth(), 0f);
                setDrawerOffset(targetOffset);
                long now = event.getEventTime();
                long elapsed = Math.max(1L, now - drawerLastEventTime);
                drawerVelocityX = (x - drawerLastX) / elapsed;
                drawerLastX = x;
                drawerLastEventTime = now;
                return true;
            case MotionEvent.ACTION_UP:
            case MotionEvent.ACTION_CANCEL:
                if (drawerDragging) {
                    finishDrawerGesture();
                    return true;
                }
                if (drawerGestureCandidate && isDrawerVisible() && x > drawerPanel.getWidth()) {
                    closeDrawer();
                    drawerGestureCandidate = false;
                    drawerGestureStartedOpen = false;
                    return true;
                }
                drawerGestureCandidate = false;
                drawerGestureStartedOpen = false;
                return false;
            default:
                return false;
        }
    }

    public void openDrawer() {
        if (drawerPanel == null) {
            return;
        }
        if (drawerPanel.getWidth() == 0) {
            drawerPanel.post(this::openDrawer);
            return;
        }
        drawerOpen = true;
        animateDrawerTo(0f, 320L);
    }

    public void closeDrawer() {
        if (drawerPanel == null || drawerPanel.getWidth() == 0) {
            return;
        }
        drawerOpen = false;
        animateDrawerTo(-drawerPanel.getWidth(), 240L);
    }

    public boolean isDrawerVisible() {
        return drawerPanel != null && (drawerPanel.getVisibility() == View.VISIBLE || drawerDragging);
    }

    public boolean consumePendingChildTouchCancel() {
        boolean shouldCancel = pendingChildTouchCancel;
        pendingChildTouchCancel = false;
        return shouldCancel;
    }

    private void updateNavigationState() {
        styleNavItem(navBookshelf, navBookshelfText, currentSection == SECTION_BOOKSHELF);
        styleNavItem(navPreview, navPreviewText, currentSection == SECTION_PREVIEW);
        styleNavItem(navSettings, navSettingsText, currentSection == SECTION_SETTINGS);
    }

    private void styleNavItem(View container, TextView textView, boolean selected) {
        if (container == null) {
            return;
        }
        container.setBackgroundResource(selected ? R.drawable.bg_nav_item_active : R.drawable.bg_nav_item_idle);
        if (textView != null) {
            textView.setTextColor(activity.getColor(selected ? R.color.primary : R.color.on_surface));
        }
    }

    private boolean shouldStartDrawerGesture(float x) {
        if (!isDrawerVisible()) {
            if (currentSection == SECTION_BOOKSHELF || currentSection == SECTION_PREVIEW) {
                return true;
            }
            return x <= closedDrawerActivationWidth();
        }
        float panelWidth = drawerPanel.getWidth();
        return x <= panelWidth + dp(OPEN_DRAWER_DRAG_EXTRA_DP);
    }

    private void finishDrawerGesture() {
        float offset = drawerPanel.getTranslationX();
        float panelWidth = Math.max(drawerPanel.getWidth(), 1);
        boolean shouldOpen;
        if (drawerGestureStartedOpen) {
            float closeThreshold = -panelWidth * CLOSE_THRESHOLD_RATIO;
            shouldOpen = !(offset <= closeThreshold || drawerVelocityX < -FLING_VELOCITY_THRESHOLD);
            if (drawerVelocityX > FLING_VELOCITY_THRESHOLD) {
                shouldOpen = true;
            }
        } else {
            float openThreshold = -panelWidth * (1f - OPEN_THRESHOLD_RATIO);
            shouldOpen = offset > openThreshold || drawerVelocityX > FLING_VELOCITY_THRESHOLD;
            if (drawerVelocityX < -FLING_VELOCITY_THRESHOLD) {
                shouldOpen = false;
            }
        }
        drawerGestureCandidate = false;
        drawerDragging = false;
        drawerGestureStartedOpen = false;
        if (shouldOpen) {
            openDrawer();
        } else {
            closeDrawer();
        }
    }

    private void animateDrawerTo(float targetOffset, long durationMs) {
        prepareDrawerForInteraction(true);
        drawerPanel.animate().cancel();
        drawerScrim.animate().cancel();
        final long animationToken = ++drawerAnimationToken;
        float startOffset = clamp(drawerPanel.getTranslationX(), -drawerPanel.getWidth(), 0f);
        float panelWidth = Math.max(drawerPanel.getWidth(), 1f);
        float remainingRatio = Math.abs(targetOffset - startOffset) / panelWidth;
        if (remainingRatio <= 0.015f) {
            applyDrawerOffsetState(targetOffset, false);
            return;
        }
        float progress = drawerProgressForOffset(targetOffset);
        float scrimAlpha = drawerScrimAlpha(progress);
        
        // Use different interpolators for open and close actions
        boolean isOpening = targetOffset == 0f;
        android.view.animation.Interpolator interpolator = isOpening 
                ? new DecelerateInterpolator(1.3f) 
                : new android.view.animation.AccelerateInterpolator(1.2f);

        drawerPanel.animate()
                .translationX(targetOffset)
                .alpha(drawerPanelAlpha(progress))
                .scaleY(drawerPanelScaleY(progress))
                .setDuration(durationMs)
                .setInterpolator(interpolator)
                .withEndAction(() -> {
                    if (animationToken != drawerAnimationToken) {
                        return;
                    }
                    applyDrawerOffsetState(targetOffset, false);
                })
                .start();
        drawerScrim.animate()
                .alpha(scrimAlpha)
                .setDuration(durationMs)
                .setInterpolator(isOpening ? new DecelerateInterpolator(1.3f) : new android.view.animation.AccelerateInterpolator(1.2f))
                .start();
    }

    private void prepareDrawerForInteraction(boolean forceVisible) {
        drawerAnimationToken++;
        drawerScrim.bringToFront();
        drawerPanel.bringToFront();
        drawerPanel.animate().cancel();
        drawerScrim.animate().cancel();
        float panelWidth = Math.max(drawerPanel.getWidth(), 1);
        float currentOffset = clamp(drawerPanel.getTranslationX(), -panelWidth, 0f);
        applyDrawerOffsetState(currentOffset, forceVisible);
    }

    private void setDrawerOffset(float offset) {
        float clamped = clamp(offset, -drawerPanel.getWidth(), 0f);
        applyDrawerOffsetState(clamped, false);
    }

    private float clamp(float value, float min, float max) {
        return Math.max(min, Math.min(max, value));
    }

    private int dp(int value) {
        return Math.round(activity.getResources().getDisplayMetrics().density * value);
    }

    private int closedDrawerActivationWidth() {
        int panelWidth = drawerPanel == null ? 0 : drawerPanel.getWidth();
        int preferred = panelWidth > 0 ? Math.round(panelWidth * 0.56f) : dp(EDGE_ACTIVATION_MIN_DP);
        return clamp(preferred, dp(EDGE_ACTIVATION_MIN_DP), dp(EDGE_ACTIVATION_MAX_DP));
    }

    private int clamp(int value, int min, int max) {
        return Math.max(min, Math.min(max, value));
    }

    private void applyDrawerOffsetState(float offset, boolean keepVisible) {
        if (drawerPanel == null || drawerScrim == null) {
            return;
        }
        float panelWidth = Math.max(drawerPanel.getWidth(), 1f);
        float clampedOffset = clamp(offset, -panelWidth, 0f);
        float progress = drawerProgressForOffset(clampedOffset);
        float scrimAlpha = drawerScrimAlpha(progress);
        boolean showPanel = keepVisible || progress > 0.001f;
        boolean showScrim = keepVisible || scrimAlpha > 0f;

        drawerPanel.setPivotX(0f);
        drawerPanel.setPivotY(drawerPanel.getHeight() * 0.5f);
        drawerPanel.setVisibility(showPanel ? View.VISIBLE : View.INVISIBLE);
        drawerPanel.setTranslationX(clampedOffset);
        drawerPanel.setAlpha(showPanel ? drawerPanelAlpha(progress) : 0f);
        drawerPanel.setScaleY(drawerPanelScaleY(progress));

        drawerScrim.setVisibility(showScrim ? View.VISIBLE : View.GONE);
        drawerScrim.setAlpha(showScrim ? scrimAlpha : 0f);
        drawerOpen = progress > 0.95f;
    }

    private float drawerProgressForOffset(float offset) {
        float panelWidth = Math.max(drawerPanel.getWidth(), 1f);
        float clampedOffset = clamp(offset, -panelWidth, 0f);
        return 1f - (-clampedOffset / panelWidth);
    }

    private float drawerPanelAlpha(float progress) {
        float safeProgress = Math.max(0f, Math.min(1f, progress));
        if (safeProgress <= 0f) {
            return 0f;
        }
        float easedProgress = 1f - (float) Math.pow(1f - safeProgress, 1.18f);
        return 0.74f + 0.26f * easedProgress;
    }

    private float drawerPanelScaleY(float progress) {
        float safeProgress = Math.max(0f, Math.min(1f, progress));
        if (safeProgress <= 0f) {
            return 0.986f;
        }
        float easedProgress = 1f - (float) Math.pow(1f - safeProgress, 1.08f);
        return 0.986f + 0.014f * easedProgress;
    }

    private float drawerScrimAlpha(float progress) {
        float safeProgress = Math.max(0f, Math.min(1f, progress));
        if (safeProgress <= 0.04f) {
            return 0f;
        }
        float easedProgress = (safeProgress - 0.04f) / 0.96f;
        easedProgress = (float) Math.pow(easedProgress, 1.35f);
        return Math.min(0.92f, easedProgress);
    }

    private void handleRootLayoutChanged() {
        updateDrawerPanelWidth();
        syncDrawerOffsetToCurrentState();
        updateDrawerGestureExclusion();
    }

    private void updateDrawerPanelWidth() {
        if (drawerPanel == null) {
            return;
        }
        int containerWidth = rootView != null ? rootView.getWidth() : 0;
        if (containerWidth <= 0) {
            return;
        }
        int desiredWidth = clamp(
                Math.round(containerWidth * DRAWER_WIDTH_RATIO),
                dp(DRAWER_MIN_WIDTH_DP),
                dp(DRAWER_MAX_WIDTH_DP)
        );
        ViewGroup.LayoutParams layoutParams = drawerPanel.getLayoutParams();
        if (layoutParams == null || layoutParams.width == desiredWidth) {
            return;
        }
        layoutParams.width = desiredWidth;
        drawerPanel.setLayoutParams(layoutParams);
    }

    private void syncDrawerOffsetToCurrentState() {
        if (drawerPanel == null || drawerPanel.getWidth() == 0) {
            return;
        }
        if (drawerDragging) {
            setDrawerOffset(drawerPanel.getTranslationX());
            return;
        }
        setDrawerOffset(drawerOpen ? 0f : -drawerPanel.getWidth());
    }

    private void updateDrawerGestureExclusion() {
        if (rootView == null || Build.VERSION.SDK_INT < Build.VERSION_CODES.Q) {
            return;
        }
        List<Rect> exclusionRects = new ArrayList<>();
        exclusionRects.add(new Rect(0, 0, Math.max(dp(EDGE_EXCLUSION_DP), closedDrawerActivationWidth()), rootView.getHeight()));
        rootView.setSystemGestureExclusionRects(exclusionRects);
    }
}

File: app/src/main/kotlin/com/metahumanz/pacilread/AppDrawerController.kt
package com.metahumanz.pacilread

import android.app.Activity
import android.graphics.Rect
import android.os.Build
import android.view.MotionEvent
import android.view.View
import android.view.ViewConfiguration
import android.view.animation.AccelerateInterpolator
import android.view.animation.DecelerateInterpolator
import android.view.animation.Interpolator
import android.widget.TextView
import com.metahumanz.pacilread.theme.ThemeModeHelper
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min
import kotlin.math.pow
import kotlin.math.roundToInt

open class AppDrawerController(
    private val activity: Activity,
    private val rootView: View?,
    private val navigationListener: NavigationListener,
) {
    fun interface NavigationListener {
        fun onDrawerDestinationSelected(destination: Int)
    }

    private val drawerPanel: View? = activity.findViewById(R.id.drawer_panel)
    private val drawerScrim: View? = activity.findViewById(R.id.drawer_scrim)
    private val navBookshelf: View? = activity.findViewById(R.id.nav_bookshelf)
    private val navPreview: View? = activity.findViewById(R.id.nav_preview)
    private val navSettings: View? = activity.findViewById(R.id.nav_settings)
    private val navBookshelfText: TextView? = activity.findViewById(R.id.text_nav_bookshelf)
    private val navPreviewText: TextView? = activity.findViewById(R.id.text_nav_preview)
    private val navSettingsText: TextView? = activity.findViewById(R.id.text_nav_settings)
    private val statusText: TextView? = activity.findViewById(R.id.text_drawer_status)
    private val touchSlop = ViewConfiguration.get(activity).scaledTouchSlop

    private var currentSection = SECTION_NONE
    private var drawerOpen = false
    private var drawerAnimating = false
    private var drawerGestureCandidate = false
    private var drawerDragging = false
    private var drawerDownX = 0f
    private var drawerDownY = 0f
    private var drawerBaseOffset = 0f
    private var drawerLastX = 0f
    private var drawerLastEventTime = 0L
    private var drawerVelocityX = 0f
    private var pendingChildTouchCancel = false
    private var drawerGestureStartedOpen = false
    private var drawerAnimationToken = 0L

    init {
        setupViews()
    }

    private fun setupViews() {
        val panel = drawerPanel ?: return
        val scrim = drawerScrim ?: return
        panel.translationX = -9999f
        panel.visibility = View.INVISIBLE
        scrim.visibility = View.GONE
        scrim.alpha = 0f
        scrim.setOnClickListener { closeDrawer() }
        navBookshelf!!.setOnClickListener {
            navigationListener.onDrawerDestinationSelected(SECTION_BOOKSHELF)
            closeDrawer()
        }
        navPreview!!.setOnClickListener {
            navigationListener.onDrawerDestinationSelected(SECTION_PREVIEW)
            closeDrawer()
        }
        navSettings!!.setOnClickListener {
            navigationListener.onDrawerDestinationSelected(SECTION_SETTINGS)
            closeDrawer()
        }
        panel.post {
            updateDrawerPanelWidth()
            setDrawerOffset(-panel.width.toFloat())
            panel.visibility = View.INVISIBLE
            scrim.visibility = View.GONE
            scrim.alpha = 0f
        }
        rootView?.let { root ->
            root.post {
                updateDrawerPanelWidth()
                syncDrawerOffsetToCurrentState()
                updateDrawerGestureExclusion()
            }
            root.addOnLayoutChangeListener { _, _, _, _, _, _, _, _, _ -> handleRootLayoutChanged() }
        }
        updateNavigationState()
    }

    fun bindMenuButton(viewId: Int) {
        activity.findViewById<View>(viewId)?.setOnClickListener { openDrawer() }
    }

    fun setCurrentSection(section: Int) {
        currentSection = section
        updateNavigationState()
    }

    fun setStatusText(text: String?) {
        statusText?.text = text.orEmpty()
    }

    fun onBackPressed(): Boolean {
        if (!isDrawerVisible()) return false
        closeDrawer()
        return true
    }

    fun handleTouchEvent(event: MotionEvent): Boolean {
        val panel = drawerPanel ?: return false
        if (panel.width == 0) return false
        val x = event.x
        val y = event.y
        return when (event.actionMasked) {
            MotionEvent.ACTION_DOWN -> {
                pendingChildTouchCancel = false
                drawerGestureCandidate = shouldStartDrawerGesture(x)
                drawerDragging = false
                drawerVelocityX = 0f
                drawerGestureStartedOpen = isDrawerVisible()
                if (drawerGestureCandidate) {
                    drawerDownX = x
                    drawerDownY = y
                    drawerLastX = x
                    drawerLastEventTime = event.eventTime
                    drawerBaseOffset = panel.translationX
                }
                false
            }

            MotionEvent.ACTION_MOVE -> {
                if (!drawerGestureCandidate) return false
                val deltaX = x - drawerDownX
                val deltaY = y - drawerDownY
                if (!drawerDragging) {
                    val horizontalThreshold = if (isDrawerVisible()) touchSlop else max(4, touchSlop / 2)
                    if (abs(deltaY) > touchSlop && abs(deltaY) > abs(deltaX)) {
                        drawerGestureCandidate = false
                        return false
                    }
                    if (abs(deltaX) <= horizontalThreshold || abs(deltaX) <= abs(deltaY)) return false
                    if (!isDrawerVisible() && deltaX < 0f) {
                        drawerGestureCandidate = false
                        return false
                    }
                    drawerDragging = true
                    pendingChildTouchCancel = true
                    prepareDrawerForInteraction(false)
                }
                setDrawerOffset(clamp(drawerBaseOffset + deltaX, -panel.width.toFloat(), 0f))
                val now = event.eventTime
                val elapsed = max(1L, now - drawerLastEventTime)
                drawerVelocityX = (x - drawerLastX) / elapsed
                drawerLastX = x
                drawerLastEventTime = now
                true
            }

            MotionEvent.ACTION_UP, MotionEvent.ACTION_CANCEL -> {
                if (drawerDragging) {
                    finishDrawerGesture()
                    true
                } else if (drawerGestureCandidate && isDrawerVisible() && x > panel.width) {
                    closeDrawer()
                    drawerGestureCandidate = false
                    drawerGestureStartedOpen = false
                    true
                } else {
                    drawerGestureCandidate = false
                    drawerGestureStartedOpen = false
                    false
                }
            }

            else -> false
        }
    }

    fun openDrawer() {
        val panel = drawerPanel ?: return
        if (panel.width == 0) {
            panel.post(::openDrawer)
            return
        }
        animateDrawerTo(0f, 320L)
    }

    fun closeDrawer() {
        val panel = drawerPanel ?: return
        if (panel.width == 0) return
        animateDrawerTo(-panel.width.toFloat(), 240L)
    }

    fun isDrawerVisible(): Boolean =
        drawerPanel != null && (drawerPanel.visibility == View.VISIBLE || drawerDragging)

    fun consumePendingChildTouchCancel(): Boolean {
        val shouldCancel = pendingChildTouchCancel
        pendingChildTouchCancel = false
        return shouldCancel
    }

    private fun updateNavigationState() {
        styleNavItem(navBookshelf, navBookshelfText, currentSection == SECTION_BOOKSHELF)
        styleNavItem(navPreview, navPreviewText, currentSection == SECTION_PREVIEW)
        styleNavItem(navSettings, navSettingsText, currentSection == SECTION_SETTINGS)
    }

    private fun styleNavItem(container: View?, textView: TextView?, selected: Boolean) {
        container ?: return
        container.setBackgroundResource(if (selected) R.drawable.bg_nav_item_active else R.drawable.bg_nav_item_idle)
        textView?.setTextColor(
            ThemeModeHelper.resolveColor(
                activity,
                if (selected) R.color.app_nav_text_active else R.color.app_nav_text_idle,
            ),
        )
    }

    private fun shouldStartDrawerGesture(x: Float): Boolean {
        if (!isDrawerVisible()) {
            if (currentSection == SECTION_BOOKSHELF || currentSection == SECTION_PREVIEW) return true
            return x <= closedDrawerActivationWidth()
        }
        return x <= drawerPanel!!.width + dp(OPEN_DRAWER_DRAG_EXTRA_DP)
    }

    private fun finishDrawerGesture() {
        val panel = drawerPanel!!
        val offset = panel.translationX
        val panelWidth = max(panel.width, 1)
        var shouldOpen: Boolean
        if (drawerGestureStartedOpen) {
            val closeThreshold = -panelWidth * CLOSE_THRESHOLD_RATIO
            shouldOpen = !(offset <= closeThreshold || drawerVelocityX < -FLING_VELOCITY_THRESHOLD)
            if (drawerVelocityX > FLING_VELOCITY_THRESHOLD) shouldOpen = true
        } else {
            val openThreshold = -panelWidth * (1f - OPEN_THRESHOLD_RATIO)
            shouldOpen = offset > openThreshold || drawerVelocityX > FLING_VELOCITY_THRESHOLD
            if (drawerVelocityX < -FLING_VELOCITY_THRESHOLD) shouldOpen = false
        }
        drawerGestureCandidate = false
        drawerDragging = false
        drawerGestureStartedOpen = false
        if (shouldOpen) openDrawer() else closeDrawer()
    }

    private fun animateDrawerTo(targetOffset: Float, durationMs: Long) {
        val panel = drawerPanel!!
        val scrim = drawerScrim!!
        panel.animate().cancel()
        scrim.animate().cancel()
        val animationToken = ++drawerAnimationToken
        drawerAnimating = true
        val startOffset = clamp(panel.translationX, -panel.width.toFloat(), 0f)
        val panelWidth = max(panel.width.toFloat(), 1f)
        val remainingRatio = abs(targetOffset - startOffset) / panelWidth
        if (remainingRatio <= 0.015f) {
            val progress = drawerProgressForOffset(targetOffset)
            drawerOpen = progress > 0.95f
            drawerAnimating = false
            applyDrawerOffsetState(targetOffset, false)
            return
        }
        val targetProgress = drawerProgressForOffset(targetOffset)
        val targetScrimAlpha = drawerScrimAlpha(targetProgress)
        val targetPanelAlpha = drawerPanelAlpha(targetProgress)
        val targetPanelScaleY = drawerPanelScaleY(targetProgress)
        val isOpening = targetOffset == 0f
        val interpolator: Interpolator = if (isOpening) DecelerateInterpolator(1.3f) else AccelerateInterpolator(1.2f)
        if (isOpening) {
            panel.visibility = View.VISIBLE
            if (scrim.visibility != View.VISIBLE) {
                scrim.visibility = View.VISIBLE
                scrim.alpha = 0f
            }
        } else {
            panel.visibility = View.VISIBLE
            if (scrim.visibility != View.VISIBLE) {
                scrim.visibility = View.VISIBLE
                scrim.alpha = drawerScrimAlpha(drawerProgressForOffset(startOffset))
            }
        }
        scrim.bringToFront()
        panel.bringToFront()
        panel.animate()
            .translationX(targetOffset)
            .alpha(targetPanelAlpha)
            .scaleY(targetPanelScaleY)
            .setDuration(durationMs)
            .setInterpolator(interpolator)
            .withEndAction {
                if (animationToken != drawerAnimationToken) return@withEndAction
                drawerOpen = targetProgress > 0.95f
                drawerAnimating = false
                if (!drawerOpen) panel.visibility = View.INVISIBLE
            }
            .start()
        scrim.animate()
            .alpha(targetScrimAlpha)
            .setDuration(durationMs)
            .setInterpolator(if (isOpening) DecelerateInterpolator(1.3f) else AccelerateInterpolator(1.2f))
            .withEndAction {
                if (animationToken != drawerAnimationToken) return@withEndAction
                if (!isOpening && targetScrimAlpha <= 0.01f) scrim.visibility = View.GONE
            }
            .start()
    }

    private fun prepareDrawerForInteraction(forceVisible: Boolean) {
        val panel = drawerPanel!!
        val scrim = drawerScrim!!
        drawerAnimationToken++
        drawerAnimating = false
        scrim.bringToFront()
        panel.bringToFront()
        panel.animate().cancel()
        scrim.animate().cancel()
        val panelWidth = max(panel.width, 1)
        val currentOffset = clamp(panel.translationX, -panelWidth.toFloat(), 0f)
        applyDrawerOffsetState(currentOffset, forceVisible)
        if (forceVisible && scrim.visibility != View.VISIBLE) {
            scrim.visibility = View.VISIBLE
            scrim.alpha = 0f
        }
    }

    private fun setDrawerOffset(offset: Float) {
        val panel = drawerPanel!!
        applyDrawerOffsetState(clamp(offset, -panel.width.toFloat(), 0f), false)
    }

    private fun clamp(value: Float, low: Float, high: Float): Float = max(low, min(high, value))

    private fun dp(value: Int): Int = (activity.resources.displayMetrics.density * value).roundToInt()

    private fun closedDrawerActivationWidth(): Int {
        val panelWidth = drawerPanel?.width ?: 0
        val preferred = if (panelWidth > 0) (panelWidth * 0.56f).roundToInt() else dp(EDGE_ACTIVATION_MIN_DP)
        return clamp(preferred, dp(EDGE_ACTIVATION_MIN_DP), dp(EDGE_ACTIVATION_MAX_DP))
    }

    private fun clamp(value: Int, low: Int, high: Int): Int = max(low, min(high, value))

    private fun applyDrawerOffsetState(offset: Float, keepVisible: Boolean) {
        val panel = drawerPanel ?: return
        val scrim = drawerScrim ?: return
        val panelWidth = max(panel.width.toFloat(), 1f)
        val clampedOffset = clamp(offset, -panelWidth, 0f)
        val progress = drawerProgressForOffset(clampedOffset)
        val scrimAlpha = drawerScrimAlpha(progress)
        val showPanel = keepVisible || progress > 0.001f
        val showScrim = keepVisible || scrimAlpha > 0f
        panel.pivotX = 0f
        panel.pivotY = panel.height * 0.5f
        panel.visibility = if (showPanel) View.VISIBLE else View.INVISIBLE
        panel.translationX = clampedOffset
        panel.alpha = if (showPanel) drawerPanelAlpha(progress) else 0f
        panel.scaleY = drawerPanelScaleY(progress)
        if (showScrim) {
            scrim.visibility = View.VISIBLE
            scrim.alpha = scrimAlpha
        } else if (scrimAlpha <= 0.01f) {
            scrim.visibility = View.GONE
            scrim.alpha = 0f
        } else {
            scrim.visibility = View.VISIBLE
            scrim.alpha = scrimAlpha
        }
        drawerOpen = progress > 0.95f
    }

    private fun drawerProgressForOffset(offset: Float): Float {
        val panelWidth = max(drawerPanel!!.width.toFloat(), 1f)
        val clampedOffset = clamp(offset, -panelWidth, 0f)
        return 1f - -clampedOffset / panelWidth
    }

    private fun drawerPanelAlpha(progress: Float): Float {
        val safeProgress = clamp(progress, 0f, 1f)
        if (safeProgress <= 0f) return 0f
        val easedProgress = 1f - (1f - safeProgress).pow(1.18f)
        return 0.74f + 0.26f * easedProgress
    }

    private fun drawerPanelScaleY(progress: Float): Float {
        val safeProgress = clamp(progress, 0f, 1f)
        if (safeProgress <= 0f) return 0.986f
        val easedProgress = 1f - (1f - safeProgress).pow(1.08f)
        return 0.986f + 0.014f * easedProgress
    }

    private fun drawerScrimAlpha(progress: Float): Float {
        val safeProgress = clamp(progress, 0f, 1f)
        if (safeProgress <= 0.04f) return 0f
        var easedProgress = (safeProgress - 0.04f) / 0.96f
        easedProgress = easedProgress.pow(1.35f)
        return min(0.92f, easedProgress)
    }

    private fun handleRootLayoutChanged() {
        updateDrawerPanelWidth()
        syncDrawerOffsetToCurrentState()
        updateDrawerGestureExclusion()
    }

    private fun updateDrawerPanelWidth() {
        val panel = drawerPanel ?: return
        val containerWidth = rootView?.width ?: 0
        if (containerWidth <= 0) return
        val desiredWidth = clamp(
            (containerWidth * DRAWER_WIDTH_RATIO).roundToInt(),
            dp(DRAWER_MIN_WIDTH_DP),
            dp(DRAWER_MAX_WIDTH_DP),
        )
        val layoutParams = panel.layoutParams
        if (layoutParams == null || layoutParams.width == desiredWidth) return
        layoutParams.width = desiredWidth
        panel.layoutParams = layoutParams
    }

    private fun syncDrawerOffsetToCurrentState() {
        val panel = drawerPanel ?: return
        if (panel.width == 0 || drawerDragging || drawerAnimating) return
        setDrawerOffset(if (drawerOpen) 0f else -panel.width.toFloat())
    }

    private fun updateDrawerGestureExclusion() {
        val root = rootView ?: return
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.Q) return
        val exclusionRects = ArrayList<Rect>(1)
        exclusionRects.add(Rect(0, 0, max(dp(EDGE_EXCLUSION_DP), closedDrawerActivationWidth()), root.height))
        root.systemGestureExclusionRects = exclusionRects
    }

    companion object {
        const val SECTION_NONE = -1
        const val SECTION_BOOKSHELF = 0
        const val SECTION_PREVIEW = 1
        const val SECTION_SETTINGS = 2
        private const val DRAWER_WIDTH_RATIO = 0.84f
        private const val DRAWER_MIN_WIDTH_DP = 300
        private const val DRAWER_MAX_WIDTH_DP = 420
        private const val EDGE_ACTIVATION_MIN_DP = 112
        private const val EDGE_ACTIVATION_MAX_DP = 156
        private const val EDGE_EXCLUSION_DP = 156
        private const val OPEN_DRAWER_DRAG_EXTRA_DP = 72
        private const val OPEN_THRESHOLD_RATIO = 0.34f
        private const val CLOSE_THRESHOLD_RATIO = 0.22f
        private const val FLING_VELOCITY_THRESHOLD = 0.55f
    }
}
